[DOCS-192] Webhooks UI (#171)
cobalthq/cobalt-product-public-docs: content/en/Integrations/webhooks.md

   Cobalt Platform-->>Your App: Finding created
   Cobalt Platform-->>Your App: Finding state changed
{{< /mermaid >}}
<br>

You can configure webhooks through the API and in the [Cobalt UI](#configure-webhooks-in-the-ui).
<!-- Provide a link to API docs or API use case. -->

## Before You Start

Before you start creating webhooks, complete the configuration in your app.

1. Generate an **authentication token** in your app. We need this token to authorize HTTP POST requests to your URL and connect to your app.
1. Make sure that the **URL** where you want to receive notifications is valid and your services work properly.

## Webhook Events

When you set up a webhook, you subscribe to the following events related to your pentests.

| Asset | Pentest | Finding |
|---|---|---|
| <li>Asset created</li><li>Asset updated</li><li>Asset deleted</li> | <li>Pentest created</li><li>[Pentest state](/penteststates/) changed</li><li>Pentest deleted</li> | <li>Finding created</li><li>Comment created for a finding</li><li>Finding state changed</li><li>Finding deleted</li>

## Configure Webhooks in the UI

Let’s configure webhooks in the Cobalt app.

### Create a Webhook

To create a webhook:

> 1. Navigate to the **Integrations** page, and then select **Webhooks** under **Native Integrations**.

Author: ```suggestion
1. In the Cobalt app, navigate to the **Integrations** page, and then select **Webhooks** under **Native Integrations**.
```

---
title: "Get Pentest Updates with Webhooks"
linkTitle: "Webhooks"
weight: 10
description: >
  Set up webhook notifications for your organization.
---

{{% pageinfo %}}
Configure webhooks to subscribe to real-time updates for your pentests.
{{% /pageinfo %}}

> With our API-based webhooks, you can set up an integration between your app and the Cobalt platform to get notifications for pentest events. We’ll send you updates for each event to your URL through an HTTP POST request.

Reviewer: I feel like this is the only time we need to mention API (possible exception, whatever CX-I comes up with)

From a UI User PoV, I'm not sure they care that our Webhooks are based on our API.

Author: @mjang-cobalt I think our integrations are aimed at advanced users (in many cases, developers). So I think it's OK to keep this info,

Reviewer 2: Some additional info:

The webhooks use a skinny payload, containing only the public id of the entity. This is for security reasons. 

So to be able to retrieve information that actually was updated about the entity, the users will need to use e.g. our public API. 

Maybe we should add some documentation on how to retrieve information about a webhook event with the API?
cc: @eugenerevzin

Author: @inikolov89 I think adding more info about APIs is out of scope for this task since we're only describing the Webhooks UI in this guide.
I plan to add API use cases to our Product Documentation with more details about using webhook APIs. You can see an example use case for assets here: https://developer.cobalt.io/apiusecases/create_asset/

> When you work with an API, you can become aware of new data in the following ways:
> 
> - Repeatedly send requests to the same [API endpoint](/getting-started/glossary/#api-endpoint) to retrieve new information. This process is known as API polling.
> - Configure a webhook that automatically sends updates for new data to the specified URL.

Reviewer: Based on my previous comment, I might:

```suggestion
When you work with our webhooks, you can monitor progress with your pentests to:

- Use polling to retrieve new information on a regular basis.
- Configure a webhook that automatically sends updates to the specified URL.
```

I don't think we need to specify "API polling", as it's functionally no different from "regular" [polling](https://en.wikipedia.org/wiki/Polling_(computer_science))

I think "updates" and "for new data" are somewhat redundant.

Reviewer: Alternative:

```suggestion
When you work with an API, you can become aware of new data in the following ways:

- Repeatedly send requests to the same [API endpoint](/getting-started/glossary/#api-endpoint) to retrieve new information.
- Configure a webhook that automatically sends updates for new data to the specified URL.
```

{{< mermaid align="left" theme="default" >}}
sequenceDiagram
  title Webhook for pentest events
   Your App->>Cobalt Platform: Notify me of new pentest events
   Cobalt Platform-->>Your App: Pentest created
   Cobalt Platform-->>Your App: Finding created
   Cobalt Platform-->>Your App: Asset updated
   Cobalt Platform-->>Your App: Finding state changed
{{< /mermaid >}}
<br>

You can configure webhooks through the API and in the [Cobalt UI](#configure-webhooks-in-the-ui).
<!-- Provide a link to API docs or API use case. -->

## Before You Start

Before you start creating webhooks, complete the configuration in your app.

> 1. Generate an **authentication token**. We need this token to authorize HTTP POST requests to your URL and connect to your app.

Reviewer: I _think_ this is the authentication token from the Native or Partner integrations? Maybe we can be more specific?

Author: @mjang-cobalt It's an authentication token from the user's app. We use it to authorize POST requests to their webhook URL.

> 1. Make sure that the **URL** where you want to receive notifications is valid and your services work properly.

Author: ```suggestion
1. Make sure that the **URL** where you want to receive notifications is valid and your services work properly. If possible, test your connections.
```

> 1. Make sure that the **URL** where you want to receive notifications is valid and your services work properly.

Reviewer: I wonder if we can/should say something like "If possible, test your connections"

Functionally, I'm concerned about firewalls which filter certain kinds of data (or stops data that's received frequently)

## Webhook Events

When you set up a webhook, you subscribe to the following events related to your pentests.

| Asset | Pentest | Finding |
|---|---|---|
| <li>Asset created</li><li>Asset updated</li><li>Asset deleted</li> | <li>Pentest created</li><li>[Pentest state](/penteststates/) changed</li><li>Pentest deleted</li> | <li>Finding created</li><li>Comment created for a finding</li><li>Finding state changed</li><li>Finding deleted</li>

## Configure Webhooks in the UI

Let’s configure webhooks in the Cobalt app.

### Create a Webhook

To create a webhook:

> 

Reviewer: Should we include a "Sign in to Cobalt" step?

Author: @mjang-cobalt Please see my previous comment and let me know if that works.

Reviewer: You do now say "In the Cobalt app"... 
(or did I miss that earlier?)

Author: I made a commit to the file, I should have added a suggestion.

1. In the Cobalt app, navigate to the **Integrations** page, and then select **Webhooks** under **Native Integrations**.
1. On the **Webhooks** page, select **Create Webhook**.
1. In the overlay that appears, specify the following:
   - **Webhook Name**
   - **Webhook URL**: URL to which Cobalt sends HTTP POST requests for pentest [events](#webhook-events).
     - Use a unique name and URL for each webhook you create.
   - **Authentication Token**: Your [token](#before-you-start) required to authorize API requests that Cobalt sends to your URL.
1. When ready, select **Save**.
1. We send a test API request to the specified URL to validate your webhook. The webhook becomes active once the validation is complete.
   - If the validation fails, we’ll disable your webhook within 24 hours. See [Troubleshooting](#troubleshooting) for more information.

![Create a webhook in the Cobalt UI](/integrations/CreateWebhook.png "Create a webhook in the Cobalt UI")

### Manage Webhooks

Now you can manage the webhooks that you [created](#create-a-webhook).

> - To **enable** or **disable** a webhook, use the toggle under **Active**.

Author: ```suggestion
- To make a webhook **inactive**, use the toggle under **Active**.
```

- To **edit** a webhook, select the three-dot icon under **Actions**, and then select **Edit Webhook**. Update webhook parameters in the overlay that appears.
- To **delete** a webhook, select the three-dot icon under **Actions**, and then select **Delete Webhook**.

>   - You can disable a webhook without deleting it.

Author: ```suggestion
  - You can make a webhook inactive without deleting it.
```

> ## Troubleshooting

Reviewer: Thinking ahead to when we implement search, I think we need a slightly more descriptive title:

```suggestion
## Troubleshoot Webhooks
```

You can check the status of your webhooks on the **Webhooks** page. The following icons indicate that there is a problem with your webhook:

- **Yellow warning**: The webhook stopped responding to events at the specified time. We’ll disable the webhook after 48 hours of failed attempts.
- **Red error**: The webhook stopped responding, and we disabled it after 48 hours of failed attempts.

> Find solutions for common troubleshooting problems in the table below.

Reviewer: I'm "used to" the [Google Developer Style Guide](https://developers.google.com/style/tables#introductory-sentences-for-tables) practice to intro a table with a colon -- but I realize Microsoft's guide is different in this area.

```suggestion
Find solutions for common troubleshooting problems in the table below:
```

| Problem | Solution |
|---|---|
| You can’t create a webhook because the name or URL already exists | Enter a unique name and URL for your webhook |
| <li>You can’t create a webhook because the validation fails</li><li>You don’t get webhook notifications to your URL</li><li>You can't enable a webhook that we disabled or you made inactive earlier</li> | We failed to validate your webhook. Check the following:<li>Your authentication token is valid. Refresh the token or generate a new one if needed.</li><li>The URL is valid and can accept requests.</li>|

> | You can't delete a webhook | Try again or contact support |

Reviewer: One more item. (My bad for not catching this earlier.)

In two other locations ([here](https://github.com/cobalthq/cobalt-product-public-docs/blame/main/content/en/Getting%20started/Assets/asset-description.md#L110) and [here](https://github.com/cobalthq/cobalt-product-public-docs/blame/3a56912d250538dbb872e81685f18802abbba604/content/en/Getting%20started/what-to-expect.md#L58)), we've said "Contact your Customer Success Manager (CSM) or support@cobalt.io."

Maybe it's time to set up a "partial" with that statement?

Author: @mjang-cobalt Here is a minor PR for that: https://github.com/cobalthq/cobalt-product-public-docs/pull/173

   Cobalt Platform-->>Your App: Pentest created
   Cobalt Platform-->>Your App: Finding created
   Cobalt Platform-->>Your App: Pentest state changed
   Cobalt Platform-->>Your App: Finding state changed
{{< /mermaid >}}
<br>

You can configure webhooks through the [API](https://docs.cobalt.io/v2/#webhooks) and in the [Cobalt UI](#configure-webhooks-in-the-ui).
<!-- Provide a link to API docs or API use case. -->

## Before You Start

Before you start creating webhooks, complete the configuration in your app.

1. Generate an **authentication token** in your app. We need this token to authorize HTTP POST requests to your URL and connect to your app.
1. Make sure that the **URL** where you want to receive notifications is valid and your services work properly. If possible, test your connections.

Read our [Best Practices](https://docs.cobalt.io/v2/#best-practices) for more information.

## Webhook Events

When you set up a webhook, you subscribe to the following events related to your pentests:

| Pentest | Finding |
|---|---|
| <li>Pentest created</li><li>[Pentest state](/penteststates/) changed</li> | <li>Finding created</li><li>Finding state changed</li>

> For security reasons, we only post essential details about webhook events, such as their ID and type. To retrieve more information about the event, use the [Cobalt API](https://docs.cobalt.io/v2/).

Author: @inikolov89 As a workaround, I added this paragraph to explain how users can get more details about webhook events.

Author: @inikolov89 Since we're getting closer to the release, can you please approve the PR with the suggested changes? Thank you.
